<html>
<head>
    <title>JBrowse - Retrieve data outside of browser</title>

    <link rel="stylesheet" type="text/css" href="css/genome.css">
    <script type="text/javascript" src="src/dojo/dojo.js" data-dojo-config="async: 1, baseUrl: './src'"></script>
    <script type="text/javascript" src="src/JBrowse/init.js"></script>
    <style>
        pre {
            white-space: pre-wrap;
            word-wrap: break-word;
        }
    </style>

    <script>


        // Require bare bones jbrowse components without using the main browser object
        require([
                    'dojo/io-query',
                    'dojo/json',
                    'JBrowse/Browser',
                    'JBrowse/Util',
                    'JBrowse/Util/GFF3',
                    'JBrowse/Store/SeqFeature/NCList',
                    'JBrowse/Model/SimpleFeature',
                    'JBrowse/View/Export',
                    'JBrowse/View/Export/GFF3'
                ],
                function (ioQuery,
                          JSON,
                          Browser,
                          Util,
                          GFF3Util,
                          NCList,
                          SimpleFeature,
                          ExportMixin,
                          GFF3Mixin) {
                    var queryParams = ioQuery.queryToObject(window.location.search.slice(1));

                    var config = {
                        containerID: "GenomeBrowser",
                        dataRoot: queryParams.data,
                        queryParams: queryParams,
                        location: queryParams.loc,
                        updateBrowserURL: true
                    };


                    var getGFF = function (conf, region, callback) {
                        var conf = dojo.mixin({
                            storeClass: 'JBrowse/Store/SeqFeature/NCList'
                        }, conf);
                        var store = new NCList(conf);
                        store.refSeq = {};
                        store.refSeq.name = region.refseq;
                        var exp = new dojo.declare([ExportMixin, GFF3Mixin])({store: store});

                        // Fake existence of jbrowse object
                        return exp.exportRegion({ref: region.refseq, start: region.start, end: region.end}, callback);
                    };


                    // Initialize jbrowse instance in unittestmode so that it doesn't actually draw the browser to screen
                    var browser = new Browser({unitTestMode: true});


                    console.log(config.location);
                    var region = Util.parseLocString(config.location);
                    region.refseq = region.ref;
                    console.log(region);
                    console.log(config.dataRoot);

                    var prefix = 'http://agrjbrowse.s3-website-us-east-1.amazonaws.com';

//    http://agrjbrowse.s3-website-us-east-1.amazonaws.com/zfin/zebrafish/tracks/All%20Genes/1/trackData.jsonz
                    var dataMappings = {
                                'data/Danio rerio': '/zfin/zebrafish'
                                , 'data/MGI': '/MGI/mouse'
                                , 'data/Caenorhabditis elegans': 'WormBase/c_elegans_PRJNA13758'
                                , 'data/Drosophila melanogaster': 'FlyBase/fruitfly'
                                , 'data/Saccharomyces cerevisiae': 'SGD/yeast'
                            }
                    ;

//    Caenorhabditis elegans -> /home/ubuntu/AGR-jbrowse/jbrowse/data/Caenorhabditis elegans/
//├── Danio rerio -> /home/ubuntu/AGR-jbrowse/jbrowse/data/Danio rerio/
//├── Drosophila melanogaster -> /home/ubuntu/AGR-jbrowse/jbrowse/data/Drosophila melanogaster/
//├── FlyBase -> /home/ubuntu/AGR-jbrowse/jbrowse/data/FlyBase/
//├── functions.conf -> /home/ubuntu/AGR-jbrowse/jbrowse/data/functions.conf
//├── Homo sapiens -> /home/ubuntu/AGR-jbrowse/jbrowse/data/human
//├── Mus musculus -> /home/ubuntu/AGR-jbrowse/jbrowse/data/MGI/
//├── Rattus norvegicus -> /home/ubuntu/AGR-jbrowse/jbrowse/data/RGD
//├── Saccharomyces cerevisiae -> /home/ubuntu/AGR-jbrowse/jbrowse/data/Saccharomyces cerevisiae/
//├── SGD -> /home/ubuntu/AGR-jbrowse/jbrowse/data/SGD
//├── WormBase -> /home/ubuntu/AGR-jbrowse/jbrowse/data/WormBase/
//└── zfin -> /home/ubuntu/AGR-jbrowse/jbrowse/data/zfin/

                    var dataRoot = dataMappings[config.dataRoot];

                    getGFF({
                        urlTemplate: prefix + dataRoot + '/tracks/All%20Genes/{refseq}/trackData.jsonz',
                        label: 'Genes',
                        browser: browser
                    }, region, function (features) {
                        document.getElementById('gff').innerHTML = features;
                    });

                    getGFF({
                        urlTemplate: prefix + dataRoot + '/tracks/All%20Genes/{refseq}/trackData.jsonz',
                        label: 'Genes',
                        browser: browser
                    }, region, function (features) {
                        var jsonFeatures = [] ;
                        var splitFeatures = features.split('\n');

                        for(f in splitFeatures){
                            var line = splitFeatures[f];
                            if(line && line.indexOf('#')!==0){
                                jsonFeatures.push(GFF3Util.parse_feature(line));
                            }
                        }
                        document.getElementById('renderedJSON').innerHTML = JSON.stringify(jsonFeatures);
                    });

                });

    </script>

</head>
<body>
<h1>Rendered Track</h1>
<pre id="renderedTrack"></pre>
<h1>Rendered JSON</h1>
<pre id="renderedJSON"></pre>
<h1>JBrowse gff store access outside of browser</h1>
<pre id="gff"></pre>
</body>
</html> 

