---
layout: post
title: JBrowse-1.11.6 maintenance release
date: 2015-02-12
tags: ["News","Software releases"]
---

I'm pleased to announce JBrowse 1.11.6 is now available! Everybody's feedback, bug reports, pull requests, and feature requests have been invaluable for making this happen. This release includes new UI options for easily setting "Log scale" on wiggle tracks and dynamically changing strandedness on RNA-seq according to multi-segment template flags and splice site (XS) flags. It also includes bug fixes for calculating SNP positions on certain flag combinations in BAM files and for reporting the correct genotypes on certain VCF tracks. Enjoy!

*   [JBrowse-1.11.6.zip](https://jbrowse.org/releases/JBrowse-1.11.6.zip) - 4.9M
file SHA1 81dd4db6d0dcab261450dab1414e343e106ed6c4
*   [JBrowse-1.11.6-dev.zip](https://jbrowse.org/releases/JBrowse-1.11.6-dev.zip) - 26M
file SHA1 c6d6ff4f68fa3ae0180bdea9c7871f9a5a4dd1d7

## Minor improvements

*   Added the ability to customize the contents of HTMLFeatures and CanvasFeatures mouseover tooltips more extensively. Thanks to David Muller for the original bug report and Colin Diesh for the fix ([issue #480](https://github.com/gmod/jbrowse/issues/480)).
*   Added new options for BigWig files to use min/max summary values when zoomed out using `scoreType`. Thanks to Scott Cain for reporting the issue and to Colin Diesh for the fix ([issue #518](https://github.com/gmod/jbrowse/issues/518)).
*   Added a checkbox for wiggle type tracks to have log scale. Thanks to Han Lin for the pull request ([issue #502](https://github.com/gmod/jbrowse/issues/502)).
*   Added the ability to display paired-end reads in the same direction on Alignments2 tracks with the "Use reverse template" option. Thanks to Tomaz Berisa and Colin Diesh for their contributions ([issue #485](https://github.com/gmod/jbrowse/issues/485)).
*   Added the ability to parse boolean literals from the .conf files.
*   Added the ability to specify a codon table (or partial codon table) for the Sequence track.
*   Added the ability to show or hide the main menu bar using the configuration file.

## Bug fixes

*   Fixed "boolean is not a function" error when using CanvasFeatures tracks with the Segments glyph in some situations.
*   Fixed a dialog box display issue for VCF variants with "no call" specified ([issue #513](https://github.com/gmod/jbrowse/issues/513)).
*   Fixed loading files from certain filepaths with special characters. Thanks to Ben Bimber for reporting and helping diagnose the bug ([issue #508](https://github.com/gmod/jbrowse/issues/508)).
*   Fixed popup-dialog callbacks not being called on BAM Alignments tracks in 1.11.5.
*   Fixed systemwide installations of JBrowse perl modules when using Module::Build.
*   Fixed a bug when displaying SNPs on hard clipped reads. Thanks to Thon de Boer for the bug report and to Colin Diesh for the fix ([issue #516](https://github.com/gmod/jbrowse/issues/516)).
*   Fixed a bug when displaying SNPs on spliced alignments. Thanks to GitHub user 09140930 for the bug report and to Colin Diesh for the fix ([issue #523](https://github.com/gmod/jbrowse/issues/523)).
*   Fixed a bug that affected some VCF and GFF popup dialogs in some browsers, particularly Chrome 38 ([issue #522](https://github.com/gmod/jbrowse/issues/522)).
*   Fixed a bug with the incremental indexing with generate-names that associated the wrong track with the name store. Thanks to Richard Hayes for reporting this issue and to Colin Diesh for the fix ([issue #526](https://github.com/gmod/jbrowse/issues/526)).
*   Fixed the `--workdir` parameter in generate-names.pl ([issue #506](https://github.com/gmod/jbrowse/issues/506)).
*   Fixed the display of alternative alleles in the genotype for VCF pop-ups ([issue #533](https://github.com/gmod/jbrowse/issues/533)).
*   Fixed an issue where some paired-end read data wouldn't display properly if they had the same start position ([issue #521](https://github.com/gmod/jbrowse/issues/521)).
*   Output .htaccess file for generate-names.pl when using the compress option. Thanks to Sebastien Carrere for reporting the bug ([issue #541](https://github.com/gmod/jbrowse/issues/541)).
*   Fixed a small bug with being able to scroll past end of chromosome when using the `--sizes` option to prepare-refseqs.pl ([issue #535](https://github.com/gmod/jbrowse/issues/535)).
*   Fixed a small internal code inconsistency with the positioning of the vertical position line. Thanks to Anurag Priyam for the fix ([issue #545](https://github.com/gmod/jbrowse/issues/545)).
*   Fixed a small bug with some stylesheets not being able to be used for the CanvasFeatures coloring, for example, LESS stylesheets. Thanks to Anurag Priyam for reporting the issue ([issue #527](https://github.com/gmod/jbrowse/issues/527)).
*   Fixed a small bug with mouseovers on Wiggle type tracks. Thanks to Han Lin for finding and fixing this bug ([issue #503](https://github.com/gmod/jbrowse/issues/503)).
*   Removed XS tag from the strand calculation for alignments and made it an optional rendering option for BAM files with the "Use XS" option. Thanks to Kieran Raine and the pull request ([issue #473](https://github.com/gmod/jbrowse/issues/473)).
*   Added a bugfix that prevented viewing the details of haploid VCF files. Thanks to Colin Diesh for finding and fixing this bug ([issue #536](https://github.com/gmod/jbrowse/issues/536)).
*   Added a bugfix for an issue that made browsing very buggy when using private browsing mode in Safari.
*   Fixed the `shortDescription` option for the mouseover description of tracks in the Hierarchical track list ([issue #553](https://github.com/gmod/jbrowse/issues/553)).
*   Re-added filter options that were missing for SNPCoverage tracks.
